sae/blocks.go

```diff
@@ -104,74 +101,148 @@ func (vm *VM) buildBlock(
 		zap.Time("block_time", blockTime),
 	)
 
-	settleAt := unix(blockTime.Add(-params.Tau))
+	// The block time is not trusted, as it can be provided maliciously during
+	// [VM.VerifyBlock]. It is allowed for [hook.Points] to further constrain
+	// the allowed block times. However, every block MUST at least satisfy these
+	// basic sanity checks.
+	if blockUnix := blockTime.Unix(); blockUnix < saeparams.TauSeconds {
+		return nil, fmt.Errorf("%w: %d < %d", errBlockTimeUnderMinimum, blockUnix, saeparams.TauSeconds)
+	}
+	if parentTime := parent.Timestamp(); blockTime.Before(parentTime) {
+		return nil, fmt.Errorf("%w: %s < %s", errBlockTimeBeforeParent, blockTime, parentTime)
+	}
+	if maxBlockTime := vm.config.Now().Add(maxFutureBlockTime); blockTime.After(maxBlockTime) {
+		return nil, fmt.Errorf("%w: %s > %s", errBlockTimeAfterMaximum, blockTime, maxBlockTime)
+	}
+
+	settleAt := unix(blockTime.Add(-saeparams.Tau))
 	lastSettled, ok, err := blocks.LastToSettleAt(settleAt, parent)
 	if err != nil {
 		return nil, err
 	}
 	if !ok {
-		log.Warn("Execution lagging when determining last block to settle")
```

Comment:
We should know when this happens because it suggests that the gas target is too high.

Reply:
I removed this because malicious nodes can force this to happen (and generally we don't log at a default enabled level to prevent malicious log spam).

But we do have a precedent for logging errors like this _after_ filtering so only the block producer could cause a log spam... So I think this is fine as a Warn.